gh-118107: Fix zipimporter ZIP64 handling.

jsirois opened the pull request:
The code that handles too large files and offsets was missing an import
and was manipulating a tuple as if it was a list.


<!-- gh-issue-number: gh-118107 -->
* Issue: gh-118107
<!-- /gh-issue-number -->

jsirois committed:
gh-118107: Fix zipimporter ZIP64 handling.

The code that handles too large files and offsets was missing an import.

itamaro commented on Lib/zipimport.py:
```diff
@@ -517,6 +517,7 @@ def _read_directory(archive):
                             num_extra_values = (len(extra_data) - 4) // 8
                             if num_extra_values > 3:
                                 raise ZipImportError(f"can't read header extra: {archive!r}", path=archive)
+                            import struct
```
why not import it at the top with the other imports? is it an expensive import?

jsirois replied:
We crossed comments here. No, not individually. On my machine its ~300us. That said, in Pex, I have to fight to shave off the ~O(50ms) boot overhead I get from the pure stdlib imports I use there; so I'm very sensitive to adding more latency to Python application boot times; especially when the addition is only needed for semi-exotic apps that hit these size limits. As I said in the other comment though, just let me know if this should be a standard import and I'll fix.

itamaro reviewed (changes requested):
thank you for catching this and the quick patch!

could you add a test case that covers this code path please?

jsirois commented on Lib/zipimport.py:
```diff
@@ -517,6 +517,7 @@ def _read_directory(archive):
                             num_extra_values = (len(extra_data) - 4) // 8
                             if num_extra_values > 3:
                                 raise ZipImportError(f"can't read header extra: {archive!r}", path=archive)
+                            import struct
```
I added this as a lazy import but let me know if this is not considered kosher. Over in Pex, where my tests caught this bug (See #118107 description), I'm very sensitive to unnecessary imports impacting boot latency; I didn't want to add the extra O(1ms) for struct / _struct here.

itamaro replied:
making expensive imports lazy is kosher :)  but typically we'd like to see a benchmark that shows the overhead is meaningful (or even measurable)

jsirois replied:
Ok, I'll move it up to the top-level then. I'm not happy about this in Pex, but I can't do better than CPython wants to do either!

jsirois committed:
Move the struct import to the top level.

jsirois committed:
Fix whitespace to match local convention for import comments.

itamaro commented on Lib/zipimport.py:
```diff
@@ -20,6 +20,7 @@
 import _imp  # for check_hash_based_pycs
 import _io  # for open
 import marshal  # for loads
+import struct  # for ZIP64 handling
```
hmm, sorry for backtracking on this, but I think we must defer this import!
with it at module scope, the `test_cmd_line_script.CmdLineTest.test_module_in_package_in_zipfile` test is failing, while with the deferred import it's passing.
maybe there an import cycle happening? I didn't dig deeper.

jsirois replied:
I dug but got none the wiser. I did re-introduce the deferred import in the end.

jsirois committed:
Add a test for large files and fix fallout in zipimporter.

jsirois commented on Lib/test/test_zipimport.py:
```diff
@@ -810,6 +812,27 @@ def testZip64CruftAndComment(self):
         self.doTest(".py", files, "f65536", comment=b"c" * ((1 << 16) - 1))
 
+    def testZip64LargeFile(self):
+        support.requires(
```
I had introduced the decorator from `Lib/test/test_largefile.py` and then stumbled upon the resources support (`-u`) that seems more widely used; so I went with it instead. The tests only run (and pass) with `-ulargefile`:
```
./python -m test test_zipimport -ulargefile -v -m testZip64LargeFile
== CPython 3.13.0a6+ (heads/fix-issue-118107:8d55e2f226, Apr 19 2024, 20:12:18) [GCC 11.4.0]
== Linux-5.15.146.1-microsoft-standard-WSL2-x86_64-with-glibc2.35 little-endian
== Python build: debug
== cwd: /home/jsirois/dev/python/cpython/build/test_python_worker_191158æ
== CPU count: 16
== encodings: locale=UTF-8 FS=utf-8
== resources (1): largefile

Using random seed: 3331549713
0:00:00 load avg: 3.80 Run 1 test sequentially
0:00:00 load avg: 3.80 [1/1] test_zipimport
testZip64LargeFile (test.test_zipimport.CompressedZipImportTestCase.testZip64LargeFile) ... ok
testZip64LargeFile (test.test_zipimport.UncompressedZipImportTestCase.testZip64LargeFile) ... ok

----------------------------------------------------------------------
Ran 2 tests in 42.905s

OK
test_zipimport passed in 43.0 sec

== Tests result: SUCCESS ==

1 test OK.

Total duration: 43.0 sec
Total tests: run=2 (filtered)
Total test files: run=1/1 (filtered)
Result: SUCCESS
```

jsirois commented on Lib/zipimport.py:
```diff
@@ -517,8 +517,9 @@ def _read_directory(archive):
-                            values = struct.unpack_from(f"<{min(num_extra_values, 3)}Q",
-                                                        extra_data, offset=4)
+                            import struct
+                            values = list(struct.unpack_from(f"<{min(num_extra_values, 3)}Q",
```
The new test caught the need for the list here. Prior to the fix you'd see:
```
ERROR: testZip64LargeFile (test.test_zipimport.UncompressedZipImportTestCase.testZip64LargeFile)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "<frozen importlib._bootstrap_external>", line 1510, in _path_importer_cache
KeyError: '/home/jsirois/dev/python/cpython/build/test_python_worker_140690æ/junk95142.zip'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/jsirois/dev/python/cpython/Lib/test/test_zipimport.py", line 89, in wrapper
    return fun(*args, **kwargs)
           ~~~^^^^^^^^^^^^^^^^^
  File "/home/jsirois/dev/python/cpython/Lib/test/test_zipimport.py", line 89, in wrapper
    return fun(*args, **kwargs)
           ~~~^^^^^^^^^^^^^^^^^
  File "/home/jsirois/dev/python/cpython/Lib/test/test_zipimport.py", line 849, in testZip64LargeFile
    self.doTestWithPreBuiltZip(".py", "large_file")
    ~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^
  File "/home/jsirois/dev/python/cpython/Lib/test/test_zipimport.py", line 157, in doTestWithPreBuiltZip
    mod = importlib.import_module(".".join(modules))
          ~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^
  File "/home/jsirois/dev/python/cpython/Lib/importlib/__init__.py", line 88, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1387, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1322, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 1262, in _find_spec
  File "<frozen importlib._bootstrap_external>", line 1553, in find_spec
  File "<frozen importlib._bootstrap_external>", line 1525, in _get_spec
  File "<frozen importlib._bootstrap_external>", line 1512, in _path_importer_cache
  File "<frozen importlib._bootstrap_external>", line 1488, in _path_hooks
  File "<frozen zipimport>", line 98, in __init__
  File "<frozen zipimport>", line 528, in _read_directory
AttributeError: 'tuple' object has no attribute 'pop'
```

itamaro replied:
another good catch! glad you added the test case!

jsirois commented on Lib/test/test_zipimport.py:
```diff
@@ -810,6 +812,27 @@ def testZip64CruftAndComment(self):
+            z.writestr("module.py", test_src)
+            z.write(os_helper.TESTFN, "data2")
+
+        self.doTestWithPreBuiltZip(".py", "module")
```
N.B.: From the Ubuntu build and test shard:
```
2024-04-20T03:21:40.1307092Z 10 slowest tests:
2024-04-20T03:21:40.1307830Z - test_zipimport: 3 min 9 sec
2024-04-20T03:21:40.1308498Z - test.test_multiprocessing_spawn.test_processes: 1 min 23 sec
2024-04-20T03:21:40.1353913Z - test.test_multiprocessing_forkserver.test_processes: 1 min 1 sec
2024-04-20T03:21:40.1355382Z - test_free_threading: 59.5 sec
2024-04-20T03:21:40.1355983Z - test_signal: 54.2 sec
2024-04-20T03:21:40.1356469Z - test_imaplib: 52.8 sec
2024-04-20T03:21:40.1356954Z - test_socket: 52.4 sec
2024-04-20T03:21:40.1357581Z - test.test_concurrent_futures.test_wait: 49.4 sec
2024-04-20T03:21:40.1358301Z - test_fstring: 44.3 sec
2024-04-20T03:21:40.1358854Z - test_regrtest: 41.5 sec
```

Not awesome. The `Lib/test/test_zipfile64.py` uses the fake `extralargefile` resource such that the `-u` used by CI does not enable it:
```
Lib/test/test_zipfile64.py:# XXX(nnorwitz): disable this test by looking for extralargefile resource,
Lib/test/test_zipfile64.py:        'extralargefile',
```

I could do similar, but agree with the nnorwitz that that's `# XXX`!

gpshead replied:
The test execution time and excess IO and CPU in common cases is the one thing holding me back from wanting to merge this as is.  I don't think git supports sparse files itself, but checking in the data to reconstruct the zips is feasible.  Take a look at my first draft attempt to do that in https://github.com/gpshead/cpython/commit/a1af9d02c21710ea6cdaad1a7724962855699fe6 on my fork of this branch.

thoughts?

gpshead replied:
If merging that change back to the PR branch, I'd probably elide the -c8- (ZIP_DEFLATE) compressed zip file part and adjust the code not try to auto-generate parts for anything other than ZIP_STORED (-c0-) if they are not found.  The goal of this change is to prevent ridiculous amounts of IO being done on systems with normal unixy seek based sparse file support (ie: Linux in CI and buildbots at the very least).  It isn't a big deal otherwise as `-u largefile` and `-u cpu` are both options that default to off for people otherwise.

My change still feels a bit... over-complicated?  But does allow this zip64 large file zipimport test to be run regularly.

jsirois replied:
@gpshead it does seem complicated (I don't understand it) and it doesn't work:
```
:; git slog -1
commit ab826885dc91ad487f58679a3c3f98d4fc496937 (HEAD -> gpshead/fix-issue-118107)
gpg: Signature made Sat May  4 18:47:25 2024 CDT
gpg:                using RSA key A1FE765B15233EAD18FA6ABB93E55CB567B5C626
gpg: Good signature from "John Sirois (CODE SIGNING KEY) <jsirois@apache.org>" [unknown]
gpg:                 aka "John Sirois <john.sirois@gmail.com>" [unknown]
gpg:                 aka "John Sirois (https://people.apache.org/~jsirois/) <jsirois@apache.org>" [unknown]
gpg: WARNING: This key is not certified with a trusted signature!
gpg:          There is no indication that the signature belongs to the owner.
Primary key fingerprint: A1FE 765B 1523 3EAD 18FA  6ABB 93E5 5CB5 67B5 C626
Author: Gregory P. Smith [Google LLC] <greg@krypto.org>
Date:   Sat May 4 07:33:56 2024 +0000

    Speed testZip64LargeFile up by checking in test data.

    The code to generate the test data is included in the test itself and
    will run when testdata is not found.

 Lib/test/test_zipimport.py                               |  87 ++++++++++++++++++++++++++++++++++++++++++++------
 Lib/test/zipimport_data/sparse-zip64-c0-0x000000000.part | Bin 0 -> 4096 bytes
 Lib/test/zipimport_data/sparse-zip64-c0-0x100000000.part | Bin 0 -> 4096 bytes
 Lib/test/zipimport_data/sparse-zip64-c0-0x200000000.part | Bin 0 -> 546 bytes
 Lib/test/zipimport_data/sparse-zip64-c8-0x000000000.part | Bin 0 -> 8349445 bytes
 5 files changed, 77 insertions(+), 10 deletions(-)

# ~/dev/python/cpython on  gpshead/fix-issue-118107:main [$!⇕] via  v11.4.0-gcc via  v3.10.12
:; git diff
diff --git a/Lib/zipimport.py b/Lib/zipimport.py
index 4e41d10986..2b1a52c67c 100644
--- a/Lib/zipimport.py
+++ b/Lib/zipimport.py
@@ -517,9 +517,9 @@ def _read_directory(archive):
                             num_extra_values = (len(extra_data) - 4) // 8
                             if num_extra_values > 3:
                                 raise ZipImportError(f"can't read header extra: {archive!r}", path=archive)
-                            import struct
-                            values = list(struct.unpack_from(f"<{min(num_extra_values, 3)}Q",
-                                                             extra_data, offset=4))
+                            #import struct
+                            values = struct.unpack_from(f"<{min(num_extra_values, 3)}Q",
+                                                        extra_data, offset=4)

                             # N.b. Here be dragons: the ordering of these is different than
                             # the header fields, and it's really easy to get it wrong since

# ~/dev/python/cpython on  gpshead/fix-issue-118107:main [$!⇕] via  v11.4.0-gcc via  v3.10.12
:; ./python -m test test_zipimport -ulargefile -v -m testZip64LargeFile
== CPython 3.13.0a6+ (heads/fix-issue-118107:8d55e2f226, Apr 19 2024, 20:12:18) [GCC 11.4.0]
== Linux-5.15.146.1-microsoft-standard-WSL2-x86_64-with-glibc2.35 little-endian
== Python build: debug
== cwd: /home/jsirois/dev/python/cpython/build/test_python_worker_9407æ
== CPU count: 16
== encodings: locale=UTF-8 FS=utf-8
== resources (1): largefile

Using random seed: 284364736
0:00:00 load avg: 0.03 Run 1 test sequentially
0:00:00 load avg: 0.03 [1/1] test_zipimport
testZip64LargeFile (test.test_zipimport.CompressedZipImportTestCase.testZip64LargeFile) ... ok
testZip64LargeFile (test.test_zipimport.UncompressedZipImportTestCase.testZip64LargeFile) ... ok

----------------------------------------------------------------------
Ran 2 tests in 0.010s

OK

== Tests result: SUCCESS ==

1 test OK.

Total duration: 105 ms
Total tests: run=2 (filtered)
Total test files: run=1/1 (filtered)
Result: SUCCESS
```

jsirois replied:
Ah, never mind me - I forgot the need for `make` - aka Lib/ is not live. This does seem to work and the test fails as it should. I'll need to figure out what you're doing here and then add that into my branch with, perhaps, a comment to explain what the heck your code is actually doing. I won't be back at a keyboard to do this until 5/10.

jsirois replied:
Ok, got it. I was creating a sparse data file already, but a real zip. You read that real zip, recording all ~non-zero runs as data files. Since the zip is mainly zeros in the STORED case, you re-assemble these small bits of data (central directory, file info headers, etc) using seek so that the resulting zip file is itself sparse. I might knock this out here tonight before I go AFK.

jsirois replied:
OK, @gpshead, now that I understand your change I don't understand your motivation for applying the sparse zip hack only half the time given:
> The goal of this change is to prevent ridiculous amounts of IO being done on systems with normal unixy seek based sparse file support (ie: Linux in CI and buildbots at the very least). It isn't a big deal otherwise as -u largefile and -u cpu are both options that default to off for people otherwise.

I `-u largefile` and `-u cpu` are good enough to guard the ZIP_DEFLATE case, why aren't they good enough to guard the ZIP_STORED case? With your change as-is, the test drops from ~45s to ~100ms on my machine. Only using the sparse zip hack for the STORED case brings things back up to ~30s. IIUC the only thing you're worried about then in the ZIP_STORED case is the ~4GB final Zip file size ... but `-u largefile` says the person running the tests is OK with that, no?

jsirois replied:
Perhaps you're balking at the ~8MB checked in file size for the compressed zip? That would make sense to me, and, if so, I could justify the extra complexity in a comment.

jsirois committed:
Merge branch 'main' into fix-issue-118107

itamaro reviewed (approved):
thank you, this lgtm!

Eclips4 commented on Misc/NEWS.d/next/Library/2024-04-19-09-28-43.gh-issue-118107.Mdsr1J.rst:
```diff
@@ -0,0 +1,2 @@
+Fix zipimport reading of ZIP64 files with file entries that are too big or
```
```suggestion
Fix :mod:`zipimport` reading of ZIP64 files with file entries that are too big or
```

jsirois committed:
Update Misc/NEWS.d/next/Library/2024-04-19-09-28-43.gh-issue-118107.Mdsr1J.rst

Co-authored-by: Kirill Podoprigora <kirill.bast9@mail.ru>

jsirois committed:
Merge branch 'main' into fix-issue-118107

jsirois committed:
Merge branch 'main' into fix-issue-118107

gpshead committed:
Speed testZip64LargeFile up by checking in test data.

The code to generate the test data is included in the test itself and
will run when testdata is not found.

jsirois committed:
Kill ZIP_DEFLATED sparse trick and document.

jsirois committed:
Merge remote-tracking branch 'refs/remotes/upstream/main' into fix-issue-118107

encukou commented on Lib/test/test_zipimport.py:
```diff
@@ -810,6 +815,102 @@ def testZip64CruftAndComment(self):
+                                    part_fullname = os.path.join(
+                                        TEST_DATA_DIR,
+                                        "sparse-zip64-c%d-0x%09x.part" % (self.compression, offset)
+                                    )
```
Style nitpick: the line is too long.
```suggestion
                                    template = "sparse-zip64-c{}-0x{:09x}.part"
                                    part_fullname = os.path.join(
                                        TEST_DATA_DIR,
                                        template.format(
                                            self.compression, offset)
                                    )
```

jsirois replied:
Fixed. I also made the test data generation reproducible with a test of that.

encukou reviewed (approved):
The checked-in files contain mtimes, so they aren't reproducible. A few bytes were different when I re-ran the command.

I don't think this should block the PR.

jsirois committed:
Fix 80 cols and make generated test data reproducible, with added tests of test.

jsirois committed:
Merge remote-tracking branch 'upstream/main' into fix-issue-118107

ambv committed:
Merge branch 'main' into fix-issue-118107

encukou committed:
Add test/zipimport_data to Makefile
